data/transactions/logic/eval.go

```diff
@@ -1570,6 +1571,15 @@ func opSHA3_256(cx *EvalContext) error {
 	return nil
 }
 
+// Sumhash corresponds to the hash used in State Proofs
+func opSumhash(cx *EvalContext) error {
+	last := len(cx.Stack) - 1
+	h := sumhash.New512(nil)
+	h.Write(cx.Stack[last].Bytes)
+	cx.Stack[last].Bytes = h.Sum(nil)
```

Comment:
maybe `_, _ = h.Write`
`sumhash` is `go-algorand/go-sumhash`, an implementation of a hash specification (available in the repo) that (my best efforts here) calculates the 64-byte chained unsalted subset sum of the columns of a 64x128 byte matrix derived from the `crypto/sha3` SHAKE256 hash of the word "Algorand" using the top of stack as the input message.

Reply:
I am unsure if we want `_, _ = h.Write`, for the returns are length of bytes and error, while `Write` either panic or succeed in taking the bytes.

Your observation on subset sum hash is right, it is a matrix-vector multiplication, while preimage is converted into bits. Hash function's collision resilience comes from SIS hardness assumption over Z_{2^64}.

Reply:
I do wonder if I should be calling the opcode `sumhash512` though. I think I will defer to Chris Peikert on the naming.  I'll check with him.

Reply:
I only suggested `_, _` because that's what is in the usage section of `go-sumhash/README.md`
I can see `sumhash512` making sense even if it's the only one they outright specify at the moment. 
And thank you Hang for the color on the algorithm.